enclave-runtime/src/rpc/author/mod.rs

```diff
@@ -130,16 +127,38 @@ enum TopSubmissionMode {
 	SubmitWatch,
 }
 
-impl<'a, P> Author<'a, P>
+impl<TopPoolGetter> Author<TopPoolGetter>
 where
-	P: TrustedOperationPool + Sync + Send + 'static,
+	TopPoolGetter: GetTopPool + Sync + Send + 'static,
 {
+	/// Execute a specific task or function on the TOP pool
+	///
+	/// Needed to get a clean way around the mutexes in which the pool are wrapped.
+	/// Other approaches require lots of code duplication for the mutex locking and
+	/// unwrapping.
+	fn execute_on_top_pool<F, E, R>(&self, function_to_execute: F, mutex_error_func: E) -> R
+	where
+		F: FnOnce(&TopPoolGetter::TrustedOperationPool) -> R,
+		E: FnOnce(Error) -> R,
+	{
+		let pool_mutex = match self.top_pool_getter.get() {
+			Some(mutex) => mutex,
+			None => {
+				error!("Could not get mutex to trusted operation pool");
+				return (mutex_error_func)(Error::PoolError(TopPoolError::UnlockError))
+			},
+		};
+
+		let tx_pool_guard = pool_mutex.lock().unwrap();
+		(function_to_execute)(tx_pool_guard.deref())
+	}
```

Comment:
Function to unlock and unwrap the top pool mutex and execute whatever function we need. It's a bit more complex to call this now, but it allow us to have the mutex logic in one single place.